// 0g/dist/input/pointer.js
export class Pointer {
    constructor() {
        this._position = null;
        this._primaryPressed = false;
        this._primaryDown = false;
        this._primaryUp = false;
        this._secondaryPressed = false;
        this._secondaryDown = false;
        this._secondaryUp = false;
        this.handlePointerDownEvents = (ev) => {
            if (!this._position) {
                this._position = { x: ev.clientX, y: ev.clientY };
            }
            else {
                this._position.x = ev.clientX;
                this._position.y = ev.clientY;
            }
            if (ev.type === 'pointerdown') {
                if (ev.button === 0) {
                    this._primaryDown = true;
                    this._primaryPressed = true;
                }
                else if (ev.button === 2) {
                    this._secondaryDown = true;
                    this._secondaryPressed = true;
                }
            }
            else if (ev.type === 'pointerup') {
                if (ev.button === 0) {
                    this._primaryUp = true;
                    this._primaryPressed = false;
                }
                else if (ev.button === 2) {
                    this._secondaryUp = true;
                    this._secondaryPressed = false;
                }
            }
        };
        this.handlePointerMoveEvent = (ev) => {
            this._position = { x: ev.clientX, y: ev.clientY };
        };
        this.frame = () => {
            this._primaryDown = false;
            this._primaryUp = false;
            this._secondaryDown = false;
            this._secondaryUp = false;
        };
        window.addEventListener('pointerdown', this.handlePointerDownEvents);
        window.addEventListener('pointerup', this.handlePointerDownEvents);
        window.addEventListener('pointermove', this.handlePointerMoveEvent);
    }
    /** Position might be null - that means no pointer was detected */
    get position() {
        return this._position;
    }
    get primaryPressed() {
        return this._primaryPressed;
    }
    get primaryDown() {
        return this._primaryDown;
    }
    get primaryUp() {
        return this._primaryUp;
    }
    get secondaryPressed() {
        return this._secondaryPressed;
    }
    get secondaryDown() {
        return this._secondaryDown;
    }
    get secondaryUp() {
        return this._secondaryUp;
    }
}
export const pointer = new Pointer();
